File: backend/src/main/java/com/schoolfinder/security/UserResolver.java
package com.schoolfinder.security;

import com.schoolfinder.domain.AppUser;
import com.schoolfinder.domain.Role;
import com.schoolfinder.repository.AppUserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/** Maps a verified Firebase token to a persisted {@link AppUser}, creating one on first sign-in. */
@Service
public class UserResolver {

    private final AppUserRepository users;

    public UserResolver(AppUserRepository users) {
        this.users = users;
    }

    @Transactional
    public AppUser resolve(String firebaseUid, String email, String displayName) {
        AppUser existing = users.findByFirebaseUid(firebaseUid);
        if (existing != null) {
            boolean changed = false;
            if (email != null && !email.equals(existing.getEmail())) {
                existing.setEmail(email);
                changed = true;
            }
            if (displayName != null && !displayName.equals(existing.getDisplayName())) {
                existing.setDisplayName(displayName);
                changed = true;
            }
            return changed ? users.save(existing) : existing;
        }
        AppUser created = new AppUser(firebaseUid, email, displayName, Role.STUDENT);
        return users.save(created);
    }

    /** Dev-only: lets the X-Debug-Role header promote a user so admin routes can be tested. */
    @Transactional
    public void applyDevRole(AppUser user, Role role) {
        if (role != null && user.getRole() != role) {
            user.setRole(role);
        }
    }
}

File: backend/src/main/kotlin/com/schoolfinder/security/UserResolver.kt
package com.schoolfinder.security

import com.schoolfinder.domain.AppUser
import com.schoolfinder.domain.Role
import com.schoolfinder.repository.AppUserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserResolver(private val users: AppUserRepository) {

    /** Find the user for this Firebase uid, creating a STUDENT record on first sight. */
    @Transactional
    fun resolve(firebaseUid: String, email: String?, displayName: String?): AppUser {
        val existing = users.findByFirebaseUid(firebaseUid)
        if (existing != null) {
            // Keep profile fields fresh from the identity provider.
            if (email != null && existing.email != email) existing.email = email
            if (displayName != null && existing.displayName != displayName) existing.displayName = displayName
            return existing
        }
        return users.save(
            AppUser(
                firebaseUid = firebaseUid,
                email = email,
                displayName = displayName,
                role = Role.STUDENT,
            )
        )
    }

    /** Dev-only: lets the X-Debug-Role header promote a user so admin routes can be tested. */
    @Transactional
    fun applyDevRole(user: AppUser, role: Role?) {
        if (role != null && user.role != role) {
            user.role = role
        }
    }
}
